Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using BuildXL.Pips;
+using BuildXL.Pips.Operations;
+using BuildXL.Scheduler.Tracing;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Instrumentation.Common;
+
+namespace BuildXL.Scheduler.Graph
+{
+    /// <summary>
+    /// Manager which controls adding pip fragments to the graph.
+    /// </summary>
+    public class PipGraphFragmentManager : IPipGraphFragmentManager
+    {
+        private ConcurrentDictionary<string, Task<bool>> m_readFragmentTasks = new ConcurrentDictionary<string, Task<bool>>();
+
+        private IPipGraph m_pipGraph;
+
+        private PipExecutionContext m_context;
+
+        private PipGraphFragmentContext m_fragmentContext;
+
+        private LoggingContext m_loggingContext;
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContext context, IPipGraph pipGraph)
+        {
+            m_loggingContext = loggingContext;
+            m_context = context;
+            m_pipGraph = pipGraph;
+            m_fragmentContext = new PipGraphFragmentContext();
+        }
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public Task<bool> AddFragmentFileToGraph(string name, AbsolutePath filePath, string[] dependencyNames)
+        {
+            foreach (var dependencyName in dependencyNames)
+            {
+                Contract.Assert(m_readFragmentTasks.ContainsKey(dependencyName), "Fragment has not yet been added for dependency: " + dependencyName);
+            }
+
+            Task<bool> readFragmentTask = Task.Run(() =>
+            {
+                Task.WaitAll(dependencyNames.Select(dependencyName => m_readFragmentTasks[dependencyName]).ToArray());
+                return ReadPipGraphFragment(m_loggingContext, m_context, m_pipGraph, m_fragmentContext, filePath);
+            });
+
+            m_readFragmentTasks[name] = readFragmentTask;
+            return readFragmentTask;
+        }
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public async Task<bool> WaitForAllFragmentsToLoad()
+        {
+            var task = await Task.FromResult(m_readFragmentTasks.Count == 0 ? true : (await Task.WhenAll(m_readFragmentTasks.Values.ToArray())).All(x => x));
+            return task;
+        }
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public (int, int) GetPipsAdded(string name)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static bool ReadPipGraphFragment(LoggingContext loggingContext, PipExecutionContext context, IPipGraph pipGraphBuilder, PipGraphFragmentContext pipFragmentContext, AbsolutePath filePath)

```

Comment:
>ReadPipGraphFragment [](start = 28, length = 20)

We probably should have a separate PipGraphFragmentSerializer to separate out handling of serialization. It can just handle both serialization and deserialization.

Reply:
sure, located in this folder or somewhere else?

---
In reply to: [288667996](https://github.com/ljcollins25/BuildXL/pull/1#discussion_r288667996) [](ancestors = 288667996)

Reply:
In theory it could live next to RemapReader/Writer.

---
In reply to: [288691860](https://github.com/ljcollins25/BuildXL/pull/1#discussion_r288691860) [](ancestors = 288691860,288667996)

Reply:
done.

---
In reply to: [289021657](https://github.com/ljcollins25/BuildXL/pull/1#discussion_r289021657) [](ancestors = 289021657,288691860,288667996)